pkg/internal/infrastructure/helper.go

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ *
+ */
+
+package infrastructure
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+
+	networkv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+	networkclient "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/clientset/versioned/typed/k8s.cni.cncf.io/v1"
+
+	kubevirtapi "github.com/gardener/gardener-extension-provider-kubevirt/pkg/apis/kubevirt"
+	"github.com/gardener/gardener-extension-provider-kubevirt/pkg/apis/kubevirt/helper"
+
+	"github.com/gardener/gardener/pkg/utils"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	TenantNetworkAnnotation = "tenant.network.infra.gardener.io"
+)
+
+// CreateNetworkAttachmentDefinitionIfNotExist returns the GetNeworkAttachmentDefinition object matching the given GetNeworkAttachmentDefinitionRef.
+// If no existing NetworkAttachmentDefinition found, create and return one using CNI configs.
+func CreateNetworkAttachmentDefinitionIfNotExist(ctx context.Context, c client.Client, secretRef corev1.SecretReference, nadRef []*kubevirtapi.NetworkAttachmentDefinitionRef) ([]networkv1.NetworkAttachmentDefinition, error) {
+	if len(nadRef) == 0 {
+		return nil, nil
+	}
+	netClient, err := getNetworkClient(ctx, c, secretRef)
+	if err != nil {
+		return nil, err
+	}
+	nads := make([]networkv1.NetworkAttachmentDefinition, len(nadRef))
+	for i, ref := range nadRef {
+		// there could be two types of net-attach-def:
+		// pre-created ones for shared resources, e.g. storage network, that each tenant needs to attach
+		// or dynamically created ones on per tenant basis for network isolation.
+		// for pre-created network, simple get and use it. For tenant network, create it using the CNI config, and tenant network annotation.
+		nad, err := netClient.NetworkAttachmentDefinitions(ref.NetAttachDefNamespace).Get(ref.NetAttachDefName, metav1.GetOptions{})
+		if k8serrors.IsAlreadyExists(err) {
+			nads[i] = *nad
+		} else {
+			// if net-attach-def doesn't exist, create one
```

Comment:
The entire logic in this function raises a lot of concerns:
* It attempts to find existing definitions by reference. As I mention before, I don't think this is needed at all. Even if we assume it's needed, and existing NAD in a namespace different from the namespace of the VMIs is likely to be unusable, judging from Multus documentation. A reference, assuming we ever need one, is therefore just a name, the namespace must be the namespace in the provider secret.
* There should be clear distinction between NADs that we assume are preexisting and NADs we create out of configuration. The first one should only have a name (not a namespace, see above). The second one should only have a config. Please redesign the API accordingly.